blog/2024-05-21.md

```diff
@@ -0,0 +1,123 @@
+
+This way, instead of the contract calling itself on a loop waiting, the contract can simply **yield** calling the function that gives the result. When an external response is provided, the contract will **resume** and return the result.
+
+![waiting on a yield](../website/static/docs/blog/contract-wait-yield.png)
+*Contracts can now yield the execution of a function until an external service signals that the result is ready*
+
+### What is exactly being yielded?
+It is important to notice that the contract is not **halting** or **blocking** its ability to execute, nor **halting in the middle of a function** to later resume it. 
+
+In the same way that a function can return a promise to call another contract, now a function can return a **yield** to call another function.
+
+Indeed, the contract is not halting, but simply **delaying the execution of a callback** until an external agent signals that it is ok to resume.
+
+People can keep calling functions on the contract between a `yield`/`resume`, and the function that creates the `yield` can be called multiple times.
+
+:::tip
+The state **can change** between the `yield` and the `resume`, since people can keep interacting with the contract.
+
+Moreover, since the function used to signal is public, developers must make sure to guard it properly to avoid unwanted calls. This can be done by simply checking the caller of the function.
+:::
+
+### How does it change for the user?
+Between the `yield` and `resume` the user will simply be waiting to receive the result. The maximum time to wait is 200 blocks before the protocol kills the `yield` - around 4 minutes.
+
+Even better, the user does not need to expend any gas while waiting, as the contract will not be executing!.
+
+## How I can use yield/resume in my contract?
+While we have not created any official `yield/resume` example, you can refer to [Saketh Are's example](https://github.com/near/near-sdk-rs/pull/1133/files), who has been working on the `yield/resume` implementation.
+
+The basic idea is that the SDK now exposes two functions:
+- A `yield(function_to_yield)` that returns a `yield_ID` which identifies the yield
+- A `resume(yield_ID)` that signals which instance of `function_to_yield` can now execute
+
+#### Simplified Example
+
+```rust
+// const DATA_ID_REGISTER: u64 = 0;
+
+pub fn request_weather(&mut self, city: String) {
+    let index = self.next_available_request_index;
+    self.next_available_request_index += 1;
+
+    let yield_promise = env::promise_yield_create(
+        "callback_return_result",
+        &serde_json::to_vec(&(index,)).unwrap(),
+        SIGN_ON_FINISH_CALL_GAS,
+        GasWeight(0),
+        DATA_ID_REGISTER,
+    );
+
+    // Store the request, that an external service will fetch
+    let data_id: CryptoHash =
+        env::read_register(DATA_ID_REGISTER).expect("").try_into().expect("");
+    self.requests.insert(&index, WeatherRequest{&data_id, &city});
```

Comment:
This step is not strictly necessary, as the external service could observe the yield ID directly if it indexes the state of the blockchain and observes the receipts produced by this contract. If you want to show a more cut-down example in the blog post you could get rid of this part (reading and writing the data id) entirely.